packages/ra-ui-materialui/src/button/PrevNextButtons.tsx

```diff
@@ -0,0 +1,276 @@
+ * <PrevNextButtons linkType="show" />
+ *
+ * @example // share custom storeKey with other components
+ * <PrevNextButtons storeKey="listStoreKey" />
+ *
+ * @example // limit the number of records to fetch
+ * <PrevNextButtons limit={500} />
+ *
+ * @example // customize filters and sort order
+ * <PrevNextButtons
+ *     linkType="show"
+ *     sort={{
+ *         field: 'first_name',
+ *         order: 'DESC',
+ *     }}
+ *     filter={{ q: 'East a' }}
+ * />
+ *
+ * @example // customize its style
+ * <PrevNextButtons
+ *     sx={{
+ *         color: 'blue',
+ *         '& .RaPrevNextButton-list': {
+ *             marginBottom: '20px',
+ *             color: 'red',
+ *         },
+ *     }}
+ * />
+ *
+ * @example // in an edit view
+ * import * as React from "react";
+ * import { Edit, PrevNextButtons, ShowButton, SimpleForm, TopToolbar } from 'react-admin';
+ *
+ * const MyTopToolbar = ({ children }) => (
+ *     <TopToolbar>
+ *         {children}
+ *     </TopToolbar>
+ * );
+ *
+ * export const PostEdit = () => (
+ *      <Edit
+ *          actions={
+ *              <MyTopToolbar>
+ *                  <PrevNextButtons
+ *                      sort={{
+ *                          field: 'first_name',
+ *                          order: 'DESC',
+ *                      }}
+ *                      filter={{ q: 'East a' }}
+ *                  />
+ *                  <ShowButton />
+ *              </MyTopToolbar>
+ *          }
+ *      >
+ *          <SimpleForm>...</SimpleForm>
+ *      </Edit>
+ * );
+ */
+
+export const PrevNextButtons = (props: PrevNextButtonProps) => {
+    const {
+        linkType = 'edit',
+        sx,
+        storeKey,
+        limit = 1000,
+        staleTime = 5 * 60 * 1000,
+        sort = { field: 'id', order: SORT_ASC },
+        filter = {},
+    } = props;
+
+    const record = useRecordContext();
+    const resource = useResourceContext();
+
+    const [storedParams] = useStore<StoredParams>(
+        storeKey || `${resource}.listParams`,
+        {
+            filter,
+            order: sort.order,
+            sort: sort.field,
+        }
+    );
+
+    const { data, error, isLoading } = useGetList(
+        resource,
+        {
+            sort: {
+                ...{ field: storedParams.sort, order: storedParams.order },
+                ...sort,
+            },
+            filter: { ...storedParams.filter, ...filter },
+            pagination: { page: 1, perPage: limit },
+        },
+        { staleTime }
+    );
+
+    if (!record) return null;
+
+    if (isLoading) {
+        return <CircularProgress size={14} />;
+    }
+    if (error) {
+        return (
+            <ErrorIcon
+                color="error"
+                fontSize="small"
+                titleAccess="error"
+                aria-errormessage={error.message}
+            />
+        );
+    }
+
+    const ids = data ? data.map(record => record.id) : [];
+
+    const index = ids.indexOf(record.id);
```

Comment:
Unfortunately, this would not allow users to implement SaveAndGoToNext and SaveAndGoToPrev buttons.

I would extract the logic inside a `useNextPrevController` that would return an object with the following props:
- `hasNext`
- `hasPrev`
- `navigateToNext`
- `navigateToPrev`

Even better would be to also provide a `PrevNextContext` and a `PrevNextBase` component

Reply:
I followed your recommendations and extract the logic inside `useNextPrevController`. Hope it's better now. 

Next step a `PrevNextContext` and a `PrevNextBase`